core/discord/src/main/kotlin/kord/presence/KordPresenceScope.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2022 AndroidDev Discord Dev Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kord.presence
+
+import dev.kord.gateway.Gateway
+import dev.kord.gateway.start
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import presence.PresenceScope
+import presence.PresenceStatus
+
+class KordPresenceScope(
+    private val token: String,
+    private val gateway: Gateway,
+    private val scope: CoroutineScope,
+) : PresenceScope {
+    override fun updatePresence(afk: Boolean, status: PresenceStatus) {
+        scope.launch {
+            gateway.stop()
+            gateway.start(token) {
+                presence {
+                    this.afk = afk
+                    this.status = when (status) {
+                        PresenceStatus.INVISIBLE -> dev.kord.common.entity.PresenceStatus.Invisible
+                        PresenceStatus.DO_NOT_DISTURB -> dev.kord.common.entity.PresenceStatus.DoNotDisturb
+                        PresenceStatus.IDLE -> dev.kord.common.entity.PresenceStatus.Idle
+                        PresenceStatus.ONLINE -> dev.kord.common.entity.PresenceStatus.Online
+                    }
+                }
+            }
+        }
+    }
```

Comment:
why execute it as a side-effect coroutine instead of a suspend function?

Reply:
I can't see on mobile but I'm guessing this is for the `launch` that collects the SharedFlow.

I don't think it makes sense to suspend execution indefinitely when we register a command, which would require us to `launch` the register call every time for a functional bot.

Reply:
```kotlin
class KordPresenceScope(
    private val token: String,
    private val gateway: Gateway,
    private val scope: CoroutineScope,
) : PresenceScope {
    override fun updatePresence(afk: Boolean, status: PresenceStatus) {
        scope.launch {
            gateway.stop()
            gateway.start(token) {
                presence {
                    this.afk = afk
                    this.status = when (status) {
                        PresenceStatus.INVISIBLE -> dev.kord.common.entity.PresenceStatus.Invisible
                        PresenceStatus.DO_NOT_DISTURB -> dev.kord.common.entity.PresenceStatus.DoNotDisturb
                        PresenceStatus.IDLE -> dev.kord.common.entity.PresenceStatus.Idle
                        PresenceStatus.ONLINE -> dev.kord.common.entity.PresenceStatus.Online
                    }
                }
            }
        }
    }
```
the issue I have with this is that it breaks multiple boundaries
start/stop gateway and side effect

when i'm in a situation like this I prefer to have a single component in charge of receiving changes and another that can push them to it in a queue of some sort.

so you could have a single coroutine listening to this queue and performing changes

I don't want to block development over this tho'